### PR TITLE
Fixed #24391 -- Made BoundField.value() cache callable values.

### PR DESCRIPTION
**Pull request by timgraham**

### CODE REVIEW

File: tests/forms_tests/tests/test_forms.py [hunk 1]
```diff
@@ -1369,6 +1370,24 @@ class UserRegistration(Form):
...
+
+        form = MyForm()
+        name = form['name']
+        val1 = name.value()
+        val2 = name.value()
+        self.assertEqual(val1, val2)
+        # BoundField itself is also cached
+        name = form['name']
+        val3 = name.value()
+        self.assertEqual(val2, val3)
```
Comment (MarkusH) [inline tests/forms_tests/tests/test_forms.py]: Shouldn't that be `self.assertIs(name, form['name'])` to check for the same instance. No need for `val3` then.

**Comment by MarkusH** [comment]
LGTM

File: django/forms/forms.py [hunk 2]
```diff
@@ -580,12 +586,16 @@ def value(self):
         if not self.form.is_bound:
             data = self.form.initial.get(self.name, self.field.initial)
             if callable(data):
-                data = data()
-                # If this is an auto-generated default date, nix the
-                # microseconds for standardized handling. See #22502.
-                if (isinstance(data, (datetime.datetime, datetime.time)) and
-                        not getattr(self.field.widget, 'supports_microseconds', True)):
-                    data = data.replace(microsecond=0)
+                if self._initial_value is not UNSET:
```
Comment (orokusaki) [inline django/forms/forms.py]: `hasattr` is a more idiomatic approach to this (also saves a couple lines of code), unless there is a specific reason for using an object like this.

**Comment by orokusaki** [comment]
FWIW, LGTM also, with the one change Markus noted in the test. My comment is more or less a nicety.

File: django/forms/forms.py (hunk 2 above)
Comment (timgraham) [inline django/forms/forms.py]: I don't really mind either way but discussed briefly in IRC and concluded: "Code is clearer when objects of a given type always have the same set of attributes - the presence of a sentinel value is a clearer marker of intent (as opposed to bug) than the absence of an attribute entirely." -Carl Meyer

Comment (orokusaki) [inline django/forms/forms.py]: Ah, I see. I was curious whether that was a common style in Django, etc. Thanks for the reply @timgraham

**Pull request merged**